src/network/netdev/wireguard.c

```diff
@@ -0,0 +1,629 @@
+/***
+    This file is part of systemd.
+
+    Copyright 2016 Jörg Thalheim <joerg@higgsboson.tk>
+
+    systemd is free software; you can redistribute it and/or modify it
+    under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.
+
+    systemd is distributed in the hope that it will be useful, but
+    WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <sys/ioctl.h>
+#include <net/if.h>
+
+#include "alloc-util.h"
+#include "parse-util.h"
+#include "fd-util.h"
+#include "strv.h"
+#include "hexdecoct.h"
+#include "string-util.h"
+#include "wireguard.h"
+#include "networkd-link.h"
+#include "networkd-util.h"
+#include "networkd-manager.h"
+
+#define WG_SET_DEVICE (SIOCDEVPRIVATE + 1)
+#define MAX_CONCURRENT_QUERIES 2048
+
+static void resolve_endpoints(NetDev *netdev);
+
+static WireguardPeer *wireguard_peer_new(Wireguard *w, unsigned section) {
+        WireguardPeer *peer;
+
+        assert(w);
+
+        if (w->last_peer_section == section && w->peers)
+                return w->peers;
+
+        peer = new0(WireguardPeer, 1);
+        if (!peer)
+                return NULL;
+        // TODO:
+        // this should be the default, but is currently not possible as
+        // WGPEER_REMOVE_PRESHARED_KEY is not allowed for new peers
+        //peer->fields.flags = WGPEER_REPLACE_IPMASKS | WGPEER_REMOVE_PRESHARED_KEY;
```

Comment:
That's almost fair point, but not quite; see below. The semantics of this will change anyway with the move to Netlink, which upstream requires.

But one small question -- how do you handle removing peers? Looking at the source, you always set `WGDEVICE_REPLACE_PEERS`, so that the peers you do provide will replace the existing ones. That makes sense to achieve your objective. However, in that case, you don't actually need to use `WGPEER_REMOVE_PRESHARED_KEY` or any of the other REPLACE/REMOVE flags.